contracts/exchange/contracts/test/TestValidatorWallet.sol

```diff
@@ -206,41 +189,39 @@ contract TestValidatorWallet is
     }
 
     function _decodeAndValidateHashFromEncodedData(bytes memory data)
-            private
-            view
-            returns (bytes32 hash)
+        private
+        view
+        returns (bytes32 hash)
     {
-        // First we want the hash, which is the second
-        // encoded parameter. We will initially treat all fields as inline
-        // `bytes32`s and ignore the first one to extract it.
-        (, hash) = abi.decode(data, (bytes32, bytes32));
-        // Now we can figure out what the data type is from a previous call to
-        // `prepare()`.
-        DataType dataType = _hashDataTypes[hash];
-        require(
-            dataType != DataType.None,
-            "EXPECTED_NO_DATA_TYPE"
-        );
-        if (dataType == DataType.Order) {
-            // Decode the first parameter as an `Order` type.
-            LibOrder.Order memory order = abi.decode(data, (LibOrder.Order));
+        bytes4 dataId = data.readBytes4(0);
+        if (dataId == IEIP1271Data(address(0)).OrderWithHash.selector) {
+            // Decode the order and hash
+            LibOrder.Order memory order;
+            (order, hash) = abi.decode(
+                data.slice(4, data.length - 1),
```

Comment:
Why do we shave off an extra byte from the end here?

Reply:
Wait, actually, should this be `data.length - 4`? I guess it doesn't really matter.

Reply:
Whoops, off by 1 error. This should be `data.length` (it's start/end indexes, not start index + length).